### PR TITLE
[checkpoint] Implement interop between Tensor and Sharded Tensor

### PR DESCRIPTION
**Pull request by kumpera**
This allows loading a Tensor from a checkpoint with a SharedTensor in the same FQN.

### CODE REVIEW

File: torch/distributed/_shard/checkpoint/metadata.py [hunk 1]
```diff
@@ -16,8 +16,6 @@ class ShardStorageMetadata:
     shard_metadata: ShardMetadata
     # storage key used for this particular Shard
     storage_key: str
-    # Length in bytes for this shard
-    length: int
```
Comment (pritamdamania87) [inline torch/distributed/_shard/checkpoint/metadata.py:20]: Oh did we miss this redundant field earlier because it is already covered by ShardMetadata?

File: test/distributed/_shard/checkpoint/test_checkpoint.py [hunk 2]
```diff
@@ -225,7 +225,6 @@ def test_tensor_metadata_with_missing_rank_spec(self) -> None:
         (_, md) = _prepare_sharded_tensor_write(st, "tensor", mapping)
 
         self.assertEqual(1, len(md.storage_metadata))
-        self.assertEqual(4 * 4 * 8, md.storage_metadata[0].length)
```
Comment (pritamdamania87) [inline test/distributed/_shard/checkpoint/test_checkpoint.py:228]: What is the reason to remove this assertion?

File: test/distributed/_shard/checkpoint/test_file_system_checkpoint.py [hunk 3]
```diff
@@ -462,5 +462,44 @@ def test_save_load_bytes(self) -> None:
...
+    def test_switch_between_sharded_tensor_to_tensor(self) -> None:
+        path = self.get_file_path()
+
+        spec = ChunkShardingSpec(
+            dim=0,
+            placements=[
+                "rank:0/cuda:0",
+                "rank:1/cuda:1",
+            ],
+        )
```
Comment (pritamdamania87) [inline test/distributed/_shard/checkpoint/test_file_system_checkpoint.py:478]: Can we test EnumerableSpec too?

File: torch/distributed/_shard/checkpoint/state_dict_loader.py [hunk 4]
```diff
@@ -30,6 +34,27 @@
 
 from .api import CheckpointException
 
+def _create_shard_metadata(size: torch.Size, device: str) -> ShardMetadata:
+    rank = dist.get_rank() if dist.is_initialized() else 0
+    return ShardMetadata(
+        shard_offsets=[0] * len(size),
+        shard_sizes=list(size),
+        placement=f"rank:{rank}/{device}"
```
Comment (pritamdamania87) [inline torch/distributed/_shard/checkpoint/state_dict_loader.py:42]: It seems like we don't actually need the `placement` field here and need to fill it out just because `ShardMetadata` enforces this? If so, I'm wondering if we can relax this in ShardMetadata such that this can be `None` and in that case `ShardMetadata` is just referencing metadata of a ShardedTensor that hasn't really be placed anywhere yet physically (ex: checkpoint).

File: torch/distributed/_shard/checkpoint/metadata.py (hunk 1 above)
Comment (kumpera) [reply torch/distributed/_shard/checkpoint/metadata.py:20]: It's implicit from shard size + shared tensor element size.

File: test/distributed/_shard/checkpoint/test_file_system_checkpoint.py (hunk 3 above)
Comment (kumpera) [reply test/distributed/_shard/checkpoint/test_file_system_checkpoint.py:478]: good point.

File: torch/distributed/_shard/checkpoint/state_dict_loader.py (hunk 4 above)
Comment (kumpera) [reply torch/distributed/_shard/checkpoint/state_dict_loader.py:42]: Turns out ShardMetadata validation is quite centralized so this turned out to be pretty straightforward.

Comment (kumpera) [reply torch/distributed/_shard/checkpoint/state_dict_loader.py:42]: I took the opportunity to do some minor changes to ShardMetadata and fix the typing of `placement`.

**Review by pritamdamania87 — APPROVED** [review]
Thanks for adding this, changes look great! Btw, you need to add `ciflow/trunk` label to all PRs to ensure the ShardedTensor tests run. In the CI jobs, then look for `multigpu` test and ensure it passes before you land.
